dbusmock/templates/accounts_service.py

```diff
@@ -0,0 +1,392 @@
+        'IconFile': '',
+        'InputSources': dbus.Array([], signature='a{ss}'),
+        'Language': 'C',
+        'LocalAccount': True,
+        'Location': '',
+        'Locked': False,
+        'LoginFrequency': dbus.UInt64(0),
+        'LoginHistory': dbus.Array([], signature='(xxa{sv})'),
+        'LoginTime': dbus.Int64(0),
+        'PasswordHint': 'Remember it, come on!',
+        'PasswordMode': 0,
+        'Session': 'mock-session',
+        'SessionType': 'wayland',
+        'Shell': '/bin/zsh',
+        'SystemAccount': False,
+        'XHasMessages': False,
+        'XKeyboardLayouts': dbus.Array([], signature='s'),
+        'XSession': 'mock-xsession',
+    }
+    default_props.update(overrides if overrides else {})
+    self.AddObject(path, USER_IFACE, default_props, [
+        ('SetRealName', 's', '', SetRealName),
+        ('SetEmail', 's', '', SetEmail),
+        ('SetLanguage', 's', '', SetLanguage),
+        ('SetXSession', 's', '', SetXSession),
+        ('SetSession', 's', '', SetSession),
+        ('SetSessionType', 's', '', SetSessionType),
+        ('SetLocation', 's', '', SetLocation),
+        ('SetHomeDirectory', 's', '', SetHomeDirectory),
+        ('SetShell', 's', '', SetShell),
+        ('SetIconFile', 's', '', SetIconFile),
+        ('SetLocked', 'b', '', SetLocked),
+        ('SetAccountType', 'i', '', SetAccountType),
+        ('SetPasswordMode', 'i', '', SetPasswordMode),
+        ('SetPasswordHint', 's', '', SetPasswordHint),
+        ('SetPassword', 'ss', '', SetPassword),
+        ('SetAutomaticLogin', 'b', '', SetAutomaticLogin),
+        ('GetPasswordExpirationPolicy', '',
+         'xxxxxx', GetPasswordExpirationPolicy),
+    ])
+
+    had_users = len(self.mock_users) != 0
+    had_multiple_users = len(self.mock_users) > 1
+    user = mockobject.objects[path]
+    user.password = 'Pa$$wo0rd'
+    user.properties = default_props
+    user.pwd_expiration_policy = {
+        'expiration_time': 1000,
+        'last_change_time': 1615832413,
+        'min_days_between_changes': 0,
+        'max_days_between_changes': 0,
+        'days_to_warn': 0,
+        'days_after_expiration_until_lock': 0,
+    }
+    self.mock_users[uid] = default_props
+
+    self.EmitSignal(MAIN_IFACE, 'UserAdded', 'o', [path])
+
+    if not had_users:
+        emit_properties_changed(self, MAIN_IFACE, 'HasNoUsers')
+    elif not had_multiple_users and len(self.mock_users) > 1:
+        emit_properties_changed(self, MAIN_IFACE, 'HasMultipleUsers')
+
+    return path
+
+
+@dbus.service.method(MOCK_IFACE, in_signature='', out_signature='ao')
+def ListMockUsers(self):
+    """ List the mock users that have been created """
+    return [get_user_path(uid) for uid in self.mock_users.keys()]
+
+
+@dbus.service.method(MOCK_IFACE, in_signature='s', out_signature='o')
+def AddAutoLoginUser(self, username):
+    """ Enable autologin for an user """
+    path = self.FindUserByName(username)
+    self.automatic_login_users.append(path)
```

Comment:
Indeed I now saw that accountsservice has an `ao` property `AutomaticLoginUsers`, so the mock is faithful to that -- still it surprises me that this makes sense.

However, this should certainly have some error handlng if that user is already in the list?

Reply:
I'm using a set now, so... We can error indeed if called via the mock though.